dockstore-webservice/src/main/java/io/dockstore/webservice/resources/AbstractWorkflowResource.java

```diff
@@ -355,10 +360,12 @@ protected void githubWebhookRelease(String repository, String username, String g
             // If this method doesn't throw an exception, it's a valid .dockstore.yml with at least one workflow or service.
             // It also converts a .dockstore.yml 1.1 file to a 1.2 object, if necessary.
             final DockstoreYaml12 dockstoreYaml12 = DockstoreYamlHelper.readAsDockstoreYaml12(dockstoreYml.getContent());
-            createServicesAndVersionsFromDockstoreYml(dockstoreYaml12.getService(), repository, gitReference, installationId, user, dockstoreYml);
-            createBioWorkflowsAndVersionsFromDockstoreYml(dockstoreYaml12.getWorkflows(), repository, gitReference, installationId, user, dockstoreYml, false);
-            createBioWorkflowsAndVersionsFromDockstoreYml(dockstoreYaml12.getTools(), repository, gitReference, installationId, user, dockstoreYml, true);
+            List<Pair<Workflow, WorkflowVersion>> updatedWorkflowVersionPairs = new ArrayList<>();
```

Comment:
wonder if this is overkill, if I understand correctly, you only need the results of `computeName` for the left and the version name and validations for the right.

Reply:
The objects already exist, and this allows us the flexibility to use them later in whatever way we choose.  What's the advantage of precalculating the quantities you mention, and storing those, rather than whole objects?

It's important to note that `WorkflowVersion` does not currently have a method to retrieve the parent `Workflow`, which is why I'm storing pairs.

Reply:
> The objects already exist, and this allows us the flexibility to use them later in whatever way we choose. What's the advantage of precalculating the quantities you mention, and storing those, rather than whole objects? 

It's easier to reason about methods that use a small number of small (or smaller) final variables. It's kind of like a consequence of http://www.javapractices.com/topic/TopicAction.do?Id=23

In particular, these are objects tracked by the entity manager, so they're kind of the opposite of final variables in that we have to be paranoid not to modify them otherwise they can modify the database and/or cascade.

Reply:
I understand what you are saying, but, in my mind, it doesn't outweigh the cleanliness and flexibility advantages of the proposed solution.  However, it's not a huge deal, if the consensus is to unpack the objects ahead of time, I'll create a new class that stores all of the information from each pair of `Workflow` and `WorkflowVersion` objects that we might need later, add a helper function to create an instance of it, and store those.

Reply:
I thought about this a lot and was sort of going back and forth on the tradeoffs, but I lean towards Denis' approach because:

> In particular, these are objects tracked by the entity manager, so they're kind of the opposite of final variables in that we have to be paranoid not to modify them otherwise they can modify the database and/or cascade.

A variant of this is that somebody will modify them here intentionally to fix a bug or add a feature, without realizing that the change doesn't apply to all paths. If the objects were immutable, I'd be more comfortable with it, but because they're not.

Reply:
Ok, the consensus of the leads is to propagate info about each workflow/version via an intermediate object, so I will do so.

I will note, however, that the DAOs return hot, managed objects to code that responsibly reads information from them, and refrains from modification, pretty much everywhere in our system.  In the original PR, the workflows/versions don't escape from the method they're bubbled to, and if anyone was to submit code that attempted to modify them in the brief window they are accessible, multiple reviewers would pinpoint and destroy it immediately.

Reply:
So, I made a modification that will hopefully satisfy everyone.  Now, the references to each `Workflow`/`WorkflowVersion` pair are hidden inside an instance of static inner class `WorkflowAndVersion`, so they cannot be directly accessed or modified.  The methods that retrieve information forward to the appropriate `Workflow` and `WorkflowVersion` getters.

Reply:
> I will note, however, that the DAOs return hot, managed objects to code that responsibly reads information from them, and refrains from modification, pretty much everywhere in our system 

It is a little do as I say, not as I do. In our defence, I think we probably need to kick it up a notch as the system matures and grows.